plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php

```diff
@@ -274,6 +291,18 @@ public function delete_by_id( $id ) {
 	 */
 	public function delete_by_order_id( $id ) {
 		global $wpdb;
+		// Delete related records in wc_download_log (aka ON DELETE CASCADE).
+		$wpdb->query(
+			$wpdb->prepare(
+				"DELETE FROM {$wpdb->prefix}wc_download_log
+				WHERE permission_id IN (
+				    SELECT permission_id
+				    FROM {$wpdb->prefix}woocommerce_downloadable_product_permissions
+				    WHERE order_id = %d
+				)",
+				$id
+			)
+		);
```

Comment:
Similarly, this code is repeated four times, changing only the field of the `where`. I think a `delete_download_log_by_field_value($field, $value)` method would be appropriate here.